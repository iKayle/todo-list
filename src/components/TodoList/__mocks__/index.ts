 const tasksMock = [{
    id: '1',
    title: 'Teste',
    description: 'Teste description',
    completed: true
  },
  {
    id: '2',
    title: 'Teste 2',
    description: 'Teste description 2',
    completed: false
  }
]

export default tasksMock