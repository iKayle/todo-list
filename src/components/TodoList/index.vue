<script lang="ts">
import type { TodoItemProps } from '../TodoItem/index.vue'
import { onMounted, defineComponent, reactive, computed } from 'vue'
import { RouterLink } from 'vue-router'
import TodoItem from '../TodoItem/index.vue'
import FilterButton from '../FilterButton/index.vue'

export default defineComponent({
  name: 'TodoList',
  components: {
    TodoItem,
    RouterLink,
    FilterButton
  },

  setup() {
    const taskList = reactive<TodoItemProps[]>([])

    const getTaskList = () => {
      const storedData = localStorage.getItem('taskList')
      if (storedData) {
        try {
          return JSON.parse(storedData)
        } catch (error) {
          console.error('Error parsing LocalStorage data:', error)
        }
      }
      return []
    }

    onMounted(() => {
      taskList.push(...getTaskList())
    })

    const handleDelete = (id: string) => {
      const newTaskList = taskList.filter((task: TodoItemProps) => task.id !== id)
      localStorage.setItem('taskList', JSON.stringify(newTaskList))
      taskList.splice(0, taskList.length, ...newTaskList)
    }

    const handleToggle = (id: string, completed: boolean) => {
      const taskIndex = taskList.findIndex((task: TodoItemProps) => task.id === id)
      if (taskIndex !== -1) {
        taskList[taskIndex].completed = completed

        localStorage.setItem('taskList', JSON.stringify(taskList))
      }
    }

    const filter = reactive({
      currentFilter: 'all'
    })

    const filteredTaskList = computed(() => {
      if (filter.currentFilter === 'completed') {
        return taskList.filter((task) => task.completed)
      } else if (filter.currentFilter === 'pending') {
        return taskList.filter((task) => !task.completed)
      } else {
        return taskList
      }
    })

    const setFilter = (filterType: string) => {
      filter.currentFilter = filterType
    }

    return {
      taskList,
      handleDelete,
      handleToggle,
      setFilter,
      filteredTaskList,
      filter
    }
  }
})
</script>

<template>
  <section class="p-2 container mx-auto lg:w-2/4 px-4 lg:px-0 space-y-4">
    <div class="flex justify-end">
      <button class="border p-2 rounded-md">
        <RouterLink to="/create">Create Task</RouterLink>
      </button>
    </div>
    <div class="md:space-x-4 space-x-2 flex-row flex">
      <FilterButton
        label="All"
        filterType="all"
        :setActiveFilter="setFilter"
        :isActive="filter.currentFilter === 'all'"
        :filter="filter"
      />
      <FilterButton
        label="Completed"
        filterType="completed"
        :setActiveFilter="setFilter"
        :isActive="filter.currentFilter === 'completed'"
        :filter="filter"
      />
      <FilterButton
        label="Pending"
        filterType="pending"
        :setActiveFilter="setFilter"
        :isActive="filter.currentFilter === 'pending'"
        :filter="filter"
      />
    </div>

    <div v-for="task in filteredTaskList" :key="task.id">
      <TodoItem :task="task" @delete="handleDelete" @toggle="handleToggle" :key="task.id" />
    </div>

    <div v-if="filteredTaskList.length === 0" class="flex flex-col items-center p-2">
      <img src="@/assets/emptyState.png" alt="Imagem" height="300" width="300" />
      <p class="text-lg sm:text-xl font-semibold">Nenhuma tarefa encontrada.</p>
    </div>
  </section>
</template>
