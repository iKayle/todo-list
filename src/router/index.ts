import { createRouter, createWebHistory } from 'vue-router'
import HomeView from '../views/HomeView.vue'
import CreateTodoView from '../views/CreateTodoView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'home',
      component: HomeView
    },
    {
      path: '/create',
      name: 'create',
      component: CreateTodoView
    },
    {
      path: '/edit/:taskId',
      name: 'edit',
      component: CreateTodoView,
      props: true
    }
  ]
})

export default router
