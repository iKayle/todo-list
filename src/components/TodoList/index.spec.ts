import { mount } from '@vue/test-utils'
import TodoList from './index.vue'
import router from '../../router'
import tasksMock from './__mocks__/index'


describe('TodoList', () => {
  it('renders a todo', () => {
    const wrapper = mount(TodoList, {
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.text()).toContain('Nenhuma tarefa encontrada.')
    expect(wrapper.exists()).toBe(true)
  })

  it('renders a filter buttons', () => {
    const wrapper = mount(TodoList, {
      global: {
        plugins: [router]
      }
    })

    const filterButtons = wrapper.findAll('.filter-button')
    expect(filterButtons.length).toBe(3)

    expect(filterButtons[0].text()).toContain('All')
    expect(filterButtons[1].text()).toContain('Completed')
    expect(filterButtons[2].text()).toContain('Pending')
  })

  it('renders a task correctly', async () => {
    const wrapper = mount(TodoList, {
      global: {
        plugins: [router]
      }
    })

    tasksMock.forEach((task) => {
      wrapper.vm.taskList.push(task)
    })

    await wrapper.vm.$nextTick()

    const renderedTasks = wrapper.findAll('.todo-item')

    expect(renderedTasks.length).toBeGreaterThan(0)

    tasksMock.forEach((task, index) => {
      const renderedTask = renderedTasks[index]
      expect(renderedTask.text()).toContain(task.title)
      expect(renderedTask.text()).toContain(task.description)

      const pElement = renderedTask.find('.completed')

      expect(pElement.classes('line-through')).toBe(task.completed)
      expect(pElement.classes('opacity-20')).toBe(task.completed)
    })
  })

  it('delete a task from list', async () => {
    const wrapper = mount(TodoList, {
      global: {
        plugins: [router]
      }
    })
    wrapper.vm.taskList.push(tasksMock[0])

    await wrapper.vm.$nextTick()

    const buttonDelete = await wrapper.find('.icon-trash')
    await buttonDelete.trigger('click')

    expect(wrapper.vm.taskList.length).toBe(0)
  })
})
