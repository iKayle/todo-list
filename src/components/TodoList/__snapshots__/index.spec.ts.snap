// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList > renders the component 1`] = `
"<section class=\\"p-2 container mx-auto lg:w-2/4 px-4 lg:px-0 space-y-4\\">
  <div class=\\"flex justify-end\\"><button class=\\"border p-2 rounded-md\\"><a href=\\"/create\\" class=\\"\\">Create Task</a></button></div>
  <div class=\\"md:space-x-4 space-x-2 flex-row flex\\"><button class=\\"border rounded-md p-1.5 md:p-2 lg:w-32 w-24 hover:bg-indigo-400 ease-in-out transition-all duration-75 bg-indigo-500\\" isactive=\\"true\\">All</button><button class=\\"border rounded-md p-1.5 md:p-2 lg:w-32 w-24 hover:bg-indigo-400 ease-in-out transition-all duration-75\\" isactive=\\"false\\">Completed</button><button class=\\"border rounded-md p-1.5 md:p-2 lg:w-32 w-24 hover:bg-indigo-400 ease-in-out transition-all duration-75\\" isactive=\\"false\\">Pending</button></div>
  <div class=\\"flex flex-col items-center p-2\\"><img src=\\"/src/assets/emptyState.png\\" alt=\\"Imagem\\" height=\\"300\\" width=\\"300\\">
    <p class=\\"text-lg sm:text-xl font-semibold\\">Nenhuma tarefa encontrada.</p>
  </div>
</section>"
`;
